cuda/components/reduction.cuh

```diff
@@ -119,37 +115,34 @@ namespace block {
  * as to store the return value - which is stored in the 0-th position of the
  * array.
  */
-template <size_type block_size, int32 subwarp_size, typename ValueType,
-          typename Operator>
-__device__ void reduce(ValueType *data, Operator reduce_op = Operator{})
+template <
+    typename Group, typename ValueType, typename Operator,
+    typename = xstd::enable_if_t<group::is_synchronizable_group<Group>::value>>
+__device__ void reduce(const Group &group, ValueType *data,
+                       Operator reduce_op = Operator{})
 {
-    const auto local_id = thread::get_local_thread_id<subwarp_size>();
+    const auto local_id = group.thread_rank();
 
 #pragma unroll
-    for (int k = block_size / 2; k >= cuda_config::warp_size; k /= 2) {
-        block::synchronize();
+    for (int k = group.size() / 2; k >= cuda_config::warp_size; k /= 2) {
+        group.sync();
         if (local_id < k) {
             data[local_id] = reduce_op(data[local_id], data[local_id + k]);
         }
     }
 
-    if (local_id >= cuda_config::warp_size) {
+    const auto warp = group::tiled_partition<cuda_config::warp_size>(group);
+    const auto warp_id = group.thread_rank() / warp.size();
+    if (warp_id > 0) {
```

Comment:
Just want to raise your attention here to maybe double-check this. The change to this line (from `local_id >= cuda_config::warp_size`) is not straightforward.